src/lib/comms/include/sol-coap-transport.h

```diff
@@ -0,0 +1,197 @@
+/*
+ * This file is part of the Soletta Project
+ *
+ * Copyright (C) 2016 Intel Corporation. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in
+ *     the documentation and/or other materials provided with the
+ *     distribution.
+ *   * Neither the name of Intel Corporation nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <sol-buffer.h>
+#include <sol-network.h>
+#include <sol-vector.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @file
+ * @brief Routines to handle a CoAP transport.
+ *
+ * It declares a way to create a transport for Constrained
+ * Application Protocol (CoAP).
+ */
+
+/**
+ * @defgroup CoAP_Transport CoAP_Transport
+ * @ingroup Coap
+ *
+ * @{
+ */
+
+/**
+ * @brief Struct that represents CoAP transport.
+ *
+ * It declares a required interface that CoAP uses to communicate.
+ * One can create a new transport filling this struct with the
+ * proper methods.
+ *
+ * @see sol_coap_server_new()
+ */
+struct sol_coap_transport {
```

Comment:
Well, after reviewing it couple of times I think if we change slightly the interface we can simplify it a lot.

The only difference you need in transport is actually a setup and then it's a regular socket-based operation, then if you move all of what you call transport to a way to specify your own `sol_socket`, you're good to go. Example:

```c
static struct sol_socket_type coap_socket_type; // impl should be renamed to type, remove new().
static void (*orig_socket_del)(struct sol_socket *socket);

struct sol_socket *sol_coap_new_udp_socket(const struct sol_network_link_addr *addr, bool use_dtls)
{
     struct sol_socket *socket = sol_socket_new(...);
     if (!coap_socket_type.del) {
         memcpy(&coap_socket_type, socket->type, sizeof(coap_socket_type));
         coap_socket_type.del = coap_socket_del; // overload del() so we can unregister network events
         socket->type = &coap_socket_type;
     }
     // bind
     // join mcast
     // monitor network events
     return socket;
}
```

What is missing:
 - install `sol-socket.h`
 - move `struct sol_socket_impl` parts other than `new()` to it, the `get_impl()` can then be a `sol_socket_impl_new()`

Then, when we create a BLE/Iotivity like @brunobottazzini did, we do not hack an `AF_FAMILY`, we just create a new socket manually:

```c
static const struct sol_socket_type oic_ble_socket_type = {
    .sendmsg = oic_ble_sendmsg,
    .recvmsg = oic_ble_recvmsg,
...
};

struct oic_ble_socket {
    struct sol_socket base;
};

static struct sol_socket *oic_ble_socket_new(...) {
   socket = calloc(1, sizeof(struct oic_ble_socket));
   socket->base.type = &oic_ble_socket_type;
   // ble stuff here
   return &socket->base;
}
```